Add network entitlements tests

n4ss opened the pull request:
Changes done in this PR:
- `test` rule in the Makefile fixed to include package tests in subdirectories (new tests too)
- add a seccomp default profile for test purposes in **seccomp_default.json** and a profile loader in **defaults/helpers-seccomp.go**
- add helper functions to check enabled/disabled security options in OCI spec in **defaults/helpers.go**
- add `network.*` entitlements tests in **defaults/network-ents_test.go**

Follow-up PR will contain:
- `security.*` & `host.*` entitlements tests

n4ss commented:
I'll have to rewrite `seccompSyscallWithArgsBlocked` as it doesn't handle *complex* cases such as several seccomp rule whitelisting one argument at the time.

I propose for this to:
- once:
  - go through the seccomp profile
  - build a map of [syscalls<->whitelisted/blacklisted args]
- lookup in this new map for blocked/allowed syscalls (with args)

riyazdf commented on defaults/helpers.go:
```diff
@@ -18,3 +33,188 @@ func ociProfileConversionCheck(profile secprofile.Profile, entitlementID string)
 }
+
+// FIXME factor for all defaults package and even project-wide
+func testSpec() *specs.Spec {
```
if this is only used for testing, we should move it to the test file - or is this intended to be a generic spec to be reused for defaults?

n4ss replied:
this is for test purposes only, I'll move this in a test file

riyazdf commented on defaults/helpers.go:
```diff
@@ -18,3 +33,188 @@ func ociProfileConversionCheck(profile secprofile.Profile, entitlementID string)
+}
+
+// We can't create a map to check set inclusion with specs.LinuxSeccompArg as key type in Golang so no O(n)
+func seccompArgsContains(syscallArgsSet, syscallArgsSubset []specs.LinuxSeccompArg) bool {
```
if this is only used for testing, we should consider moving it to the test file

n4ss replied:
totally, actually wanted to do that in the first place 😖 will fix

riyazdf commented on defaults/helpers.go:
```diff
@@ -18,3 +33,188 @@ func ociProfileConversionCheck(profile secprofile.Profile, entitlementID string)
+	return true
+}
+
+func matchSeccompRule(seccompRule specs.LinuxSyscall, syscallName string, syscallArgs []specs.LinuxSeccompArg) bool {
```
if this is only used for testing, we should move it to the test file

riyazdf commented on defaults/helpers.go:
```diff
@@ -18,3 +33,188 @@ func ociProfileConversionCheck(profile secprofile.Profile, entitlementID string)
+	return false
+}
+
+func seccompSyscallWithArgsBlocked(seccompProfile specs.LinuxSeccomp, syscallName types.Syscall, syscallArgs []specs.LinuxSeccompArg) bool {
```
if this is only used for testing, we should move it to the test file

riyazdf commented on defaults/helpers.go:
```diff
@@ -18,3 +33,188 @@ func ociProfileConversionCheck(profile secprofile.Profile, entitlementID string)
+	return blocked
+}
+
+func seccompSyscallsWithArgsBlocked(seccompProfile specs.LinuxSeccomp, syscallsWithArgs map[types.Syscall][]specs.LinuxSeccompArg) bool {
```
if this is only used for testing, we should move it to the test file

riyazdf commented on defaults/helpers.go:
```diff
@@ -18,3 +33,188 @@ func ociProfileConversionCheck(profile secprofile.Profile, entitlementID string)
+	return true
+}
+
+func seccompSyscallsBlocked(seccompProfile specs.LinuxSeccomp, syscallNames []types.Syscall) bool {
```
if this is only used for testing, we should move it to the test file

riyazdf commented on defaults/helpers.go:
```diff
@@ -18,3 +33,188 @@ func ociProfileConversionCheck(profile secprofile.Profile, entitlementID string)
+	return true
+}
+
+func capBlocked(linuxCaps specs.LinuxCapabilities, capability types.Capability) bool {
```
if this is only used for testing, we should move it to the test file

riyazdf commented on defaults/helpers.go:
```diff
@@ -18,3 +33,188 @@ func ociProfileConversionCheck(profile secprofile.Profile, entitlementID string)
+		capListContains(linuxCaps.Inheritable, capability) || capListContains(linuxCaps.Effective, capability))
+}
+
+func capsBlocked(linuxCaps specs.LinuxCapabilities, capabilities []types.Capability) bool {
```
if this is only used for testing, we should move it to the test file

riyazdf commented on defaults/helpers.go:
```diff
@@ -18,3 +33,188 @@ func ociProfileConversionCheck(profile secprofile.Profile, entitlementID string)
+	return true
+}
+
+func capAllowed(linuxCaps specs.LinuxCapabilities, capability types.Capability) bool {
```
if this is only used for testing, we should move it to the test file

riyazdf replied:
Same as above re: ambient caps

riyazdf commented on defaults/helpers.go:
```diff
@@ -18,3 +33,188 @@ func ociProfileConversionCheck(profile secprofile.Profile, entitlementID string)
+	return true
+}
+
+func namespaceActivated(nsList []specs.LinuxNamespace, namespace specs.LinuxNamespaceType) bool {
```
if this is only used for testing, we should move it to the test file

riyazdf commented on defaults/helpers.go:
```diff
@@ -18,3 +33,188 @@ func ociProfileConversionCheck(profile secprofile.Profile, entitlementID string)
+	return false
+}
+
+func namespacesActivated(nsList []specs.LinuxNamespace, namespaces []specs.LinuxNamespaceType) bool {
```
if this is only used for testing, we should move it to the test file

riyazdf commented on defaults/helpers.go:
```diff
@@ -18,3 +33,188 @@ func ociProfileConversionCheck(profile secprofile.Profile, entitlementID string)
+	return true
+}
+
+func namespacesDeactivated(nsList []specs.LinuxNamespace, namespaces []specs.LinuxNamespaceType) bool {
```
if this is only used for testing, we should move it to the test file

riyazdf commented on secprofile/oci-profile.go:
```diff
@@ -111,8 +111,6 @@ func allowSyscallWithArgs(seccompProfile *specs.LinuxSeccomp, syscallName types.
 
 	syscallNameToAllowStr := string(syscallName)
 
-	logrus.Errorf("Allowing syscall: %s", syscallNameToAllowStr)
```
why remove this?  It may be helpful as a debug level

n4ss replied:
I felt like this info wasn't very useful as its pretty much repeats what the function is doing. Will undo that and switch it to debug if you think we should keep it.

riyazdf replied:
ok.  Let's leave this out and create a tracking issue for logging in libentitlement.  If there is enough demand, we can add it over the core codebase all at once

n4ss replied:
Sounds good! Issue here: https://github.com/docker/libentitlement/issues/46

riyazdf commented on secprofile/oci-profile.go:
```diff
@@ -134,8 +132,6 @@ func allowSyscallWithArgs(seccompProfile *specs.LinuxSeccomp, syscallName types.
 		}
 		seccompProfile.Syscalls = append(seccompProfile.Syscalls, newRule)
-
-		logrus.Errorf("Adding allow rule for syscall: %s", syscallNameToAllowStr)
```
same comment about keeping as debug

riyazdf commented on defaults/helpers.go:
```diff
@@ -18,3 +33,188 @@ func ociProfileConversionCheck(profile secprofile.Profile, entitlementID string)
+// We can't create a map to check set inclusion with specs.LinuxSeccompArg as key type in Golang so no O(n)
+func seccompArgsContains(syscallArgsSet, syscallArgsSubset []specs.LinuxSeccompArg) bool {
+	if len(syscallArgsSet) < len(syscallArgsSubset) ||
+		(len(syscallArgsSubset) == 0 && len(syscallArgsSet) != 0) {
```
for `len(syscallArgsSubset) == 0 && len(syscallArgsSet) !=0` - isn't that a true case?  Aren't we trying to check if the `syscallArgsSet` contains all of the `syscallArgsSubset`?

Side note: might help to rename vars to better differentiate the master set vs. the subset to check for

n4ss replied:
I need to rename this function or make a wrapper around it because it's currently doing more than just checking if `syscallArgsSubset` is contained in `syscallArgsSet`.

Answer to the 1st question:
Yes in theory.. but..
No because `syscallArgSubset` contains the syscall args of the to-be-checked syscall. If there are no arguments, that means that we want to completely whitelist the syscall completely so the whitelisting rule should be absolute (hence no args either).

Not sure I'm clear there.. 
function should actually be called `seccompArgsMatched(syscallRuleArgs, syscallArgsToCheck..`

riyazdf commented on defaults/helpers.go:
```diff
@@ -18,3 +33,188 @@ func ociProfileConversionCheck(profile secprofile.Profile, entitlementID string)
+	return true
+}
+
+func capBlocked(linuxCaps specs.LinuxCapabilities, capability types.Capability) bool {
```
I'm wondering if this should be configurable to also account for ambient caps?  Docker doesn't use them as of current, but it would be useful for the future / other workloads.

n4ss replied:
[As discussed] that makes sense but for now I don't want false positive on Moby. I have to talk with the runc people to understand the tradeoffs, who might use ambient caps and why.

riyazdf commented on helpers_test.go:
```diff
@@ -0,0 +1,34 @@
+	"github.com/opencontainers/runtime-spec/specs-go"
+)
+
+func testSpec() *specs.Spec {
```
is this necessary given the same function exists in helpers.go?  We should remove one.

n4ss replied:
[As discussed] I'll move this as well as *shareable* helper code for test purposes to a `testutil` package.

riyazdf commented on defaults/network-ents_test.go:
```diff
@@ -0,0 +1,157 @@
+)
+
+func TestNetworkNoneEntitlementEnforce(t *testing.T) {
+	entitlementID := "network.none"
```
we should use the const we have for this: https://github.com/docker/libentitlement/blob/master/defaults/network-ents.go#L18

n4ss replied:
indeed, will fix for all.

riyazdf commented on defaults/network-ents_test.go:
```diff
@@ -0,0 +1,157 @@
+}
+
+func TestNetworkUserEntitlementEnforce(t *testing.T) {
+	entitlementID := "network.user"
```
same comment re: const for ID

n4ss commented on defaults/helpers-seccomp.go:
```diff
@@ -0,0 +1,24 @@
+func setDefaultSeccompProfile() (*specs.LinuxSeccomp, error) {
+	var profile specs.LinuxSeccomp
+
+	jsonProfile, err := ioutil.ReadFile("seccomp_default.json")
```
Will fix, good catch.

riyazdf commented on defaults/network-ents_test.go:
```diff
@@ -0,0 +1,157 @@
+			},
+		},
+	}
+	require.False(t, seccompSyscallsWithArgsBlocked(*newOCIProfile.OCI.Linux.Seccomp, syscallsWithArgsToAllow))
```
in a follow-up PR: it would be awesome to have an integration test that runs a container similar to how you've demoed this functionality.

n4ss replied:
This will very probably be done on the Moby side but this is the next thing on my todo-list.

n4ss commented on defaults/helpers.go:
```diff
@@ -4,8 +4,23 @@ import (
+	"reflect"
 )
 
+func capListContains(capList []string, capability types.Capability) bool {
```
Hmm capList is actually an OCI capability list.. it should already be validated I guess if filed by libentitlement.. wdyt?

riyazdf replied:
ok that's fine with me.  This is an internal helper for testing in any case.

riyazdf commented on defaults/network-ents_test.go:
```diff
@@ -0,0 +1,157 @@
+
+	require.NotNil(t, newOCIProfile.OCI.Process.Capabilities)
+
+	capsToRemove := []types.Capability{CapNetAdmin, CapNetBindService, CapNetRaw}
```
can we also test that the correct network caps are included?

riyazdf commented on defaults/network-ents_test.go:
```diff
@@ -0,0 +1,157 @@
+}
+
+func TestNetworkProxyEntitlementEnforce(t *testing.T) {
+	entitlementID := "network.proxy"
```
same comment re: const for ID

n4ss commented on defaults/helpers.go:
```diff
@@ -18,3 +33,188 @@ func ociProfileConversionCheck(profile secprofile.Profile, entitlementID string)
+
+	seccomp, err := setDefaultSeccompProfile()
+	if err != nil {
+		// In case we get an error before seccomp struct is fully updated from decoded json, we empty it manually.
```
Right, will fix.

riyazdf commented on testutils/ocispec.go:
```diff
@@ -0,0 +1,219 @@
+)
+
+var (
+	testCapslist = mobyDefaultCaps
```
I don't think this is used, should it be removed?  As discussed, it would be good to use `mobyDefaultCaps` in tests such that we check an entitlement's cap list is:

`mobyDefaultCaps + entitlement.capsToAdd - entitlement.capsToRemove`

n4ss commented:
New changes:
- moved test helpers, test profiles and utils structs to `testutils` package
- moved typed capabilities and syscalls to `osdefs` package, this will be shareable by OCI test helpers in `testutils` and actual entitlements (and tests) in `defaults` package
- check perfect matching between `test struct's capabilities + caps to add for entitlement - caps to remove for entitlement` and what we obtain when enforcing the entitlement separately with `OCICapsMatchRefWithConstraints`
- fix bugs at https://github.com/docker/libentitlement/pull/45/commits/add301075b96db1a7dd5fb0bc40c915cf6ec7e03#diff-d79a0466acbe1316a0ea9bb98db1607bR56 and https://github.com/docker/libentitlement/pull/45/commits/add301075b96db1a7dd5fb0bc40c915cf6ec7e03#diff-d79a0466acbe1316a0ea9bb98db1607bR53

ashfall commented on testutils/ocispec.go:
```diff
@@ -0,0 +1,279 @@
+}
+
+// SeccompSyscallWithArgsBlocked checks that the provided syscall and args are blocked by the seccomp profile
+func SeccompSyscallWithArgsBlocked(seccompProfile specs.LinuxSeccomp, syscallName types.Syscall, syscallArgs []specs.LinuxSeccompArg) bool {
```
naming nit: Can we rename this to indicate that this is checking something? `SeccompSyscallWithArgsBlocked` doesn't tell us if it's a checker. For example, `IsSyscallBlockedBySeccomp` or `IsSyscallWithArgsBlockedBySeccomp` is a more descriptive name.

n4ss replied:
Yeah I was wondering if the "is" was necessary and indeed it seems weird without it. Will fix all the names.

ashfall commented on testutils/ocispec.go:
```diff
@@ -0,0 +1,279 @@
+}
+
+// SeccompSyscallsWithArgsBlocked checks that the provided list of syscalls and args are blocked by the seccomp profile
+func SeccompSyscallsWithArgsBlocked(seccompProfile specs.LinuxSeccomp, syscallsWithArgs map[types.Syscall][]specs.LinuxSeccompArg) bool {
```
See above re:naming

ashfall commented on testutils/ocispec.go:
```diff
@@ -0,0 +1,279 @@
+}
+
+// SeccompSyscallsBlocked checks that the provided syscalls are blocked by the seccomp profile
+func SeccompSyscallsBlocked(seccompProfile specs.LinuxSeccomp, syscallNames []types.Syscall) bool {
```
See above re:naming

ashfall commented on testutils/ocispec.go:
```diff
@@ -0,0 +1,279 @@
+// SyscallArgsMatchSeccompRuleArgs checks that the seccomp rule's args match the provided syscall args
+func SyscallArgsMatchSeccompRuleArgs(syscallArgsRule, syscallArgsSyscall []specs.LinuxSeccompArg) bool {
+	if len(syscallArgsRule) < len(syscallArgsSyscall) ||
+		(len(syscallArgsSyscall) == 0 && len(syscallArgsRule) != 0) {
```
The `syscallArgs` in `syscallArgsSyscall` and `syscallArgsRule` seems redundant, perhaps we can rename to `argsFromSyscall` and `argsFromRule`?

n4ss replied:
Yep!

ashfall commented on testutils/ocispec.go:
```diff
@@ -0,0 +1,279 @@
+	for _, syscallArgFromSyscall := range syscallArgsSyscall {
+		found := false
+
+		for _, syscallArgsFromRule := range syscallArgsRule {
```
`syscallArgsFromRule` sounds like a list -is this intentional or should this be `syscallArgFromRule`?

n4ss replied:
Will fix!

ashfall commented on testutils/ocispec.go:
```diff
@@ -0,0 +1,279 @@
+}
+
+// NamespacesActivated checks that the namespaces in the provided ns list are enabled
+func NamespacesActivated(nsList []specs.LinuxNamespace, namespaces []specs.LinuxNamespaceType) bool {
```
naming nit: Probably `areNamespacesActivated` or `checkNamespacesDeactivation`?

ashfall commented on testutils/ocispec.go:
```diff
@@ -0,0 +1,279 @@
+}
+
+// NamespaceActivated checks that the provided namespace is enabled
+func NamespaceActivated(nsList []specs.LinuxNamespace, namespace specs.LinuxNamespaceType) bool {
```
Also, naming nit: `isNamespaceActivated`

ashfall commented on testutils/ocispec.go:
```diff
@@ -0,0 +1,279 @@
+}
+
+// NamespacesDeactivated checks that the namespaces in the provided ns list are disabled
+func NamespacesDeactivated(nsList []specs.LinuxNamespace, namespaces []specs.LinuxNamespaceType) bool {
```
nit: `areNamespacesDeactivated` or `checkNamespacesDeactivation`

n4ss commented:
> Looks like most of the test helpers are not being used outside the package, is there a reason we are defining them as exported functions? If not, let's default to unexported test helpers, and we can always export/move them if we need them elsewhere. Thoughts?

Yep, sounds good. Will fix.

> Can we decide on a convention for helpers that return a bool in that their name should indicate so? fooStatusChecker or isFooActive and so on instead of fooActivated or fooDeactivated

Good catch, `isFooActive` works for me. Will fix.

riyazdf commented:
following up on naming: @ashfall and I were discussing possible different naming conventions for the matching functions like `capsListMatchRefSet` and `OCICapsMatchRefWithConstraints`.

These functions are really checking for set equivalence, perhaps `hasEquivalentCaps` and `HasEquvalentOCICaps`? I'd also be fine with creating a tracking issue for this and punting until we have more consumers of this library give feedback.

riyazdf commented on Makefile:
```diff
@@ -18,4 +18,4 @@ lint:
 
 
 test: lint
-	go test .
```
we should add codecov to this.  Can be a separate PR / tracking issue.

n4ss replied:
will do

riyazdf commented on testutils/moby-defaults.go:
```diff
@@ -0,0 +1,77 @@
+package testutils
+
+import (
+	"encoding/json"
```
nit: can we put the system imports separately at the top?

n4ss replied:
will do, gofmt actually shuffled everything..

n4ss commented:
@riyazdf I'd prefer to do that in a follow-up PR, I'm gonna submit my second PR for the rest of the entitlement testing and then we can discuss more naming conventions. We can also discuss naming conventions non-test methods there. Sounds good?
